torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -131,7 +131,9 @@ std::vector<at::Device> getDeviceList(const std::vector<at::Tensor>& tensors) {
   std::vector<at::Device> res;
   res.reserve(tensors.size());
   for (auto& tensor : tensors) {
-    res.push_back(tensor.device());
+    if (res.size() == 0 || tensor[0].device() != res[0]) {
+      res.push_back(tensor.device());
```

Comment:
Not sure I quite understand this logic. initially res.size == 0 so we push back tensor[0]'s device. 
after that res[0] == tensor[0].device is always true. So in all iterations after the first, we don't push back anything. It seems that it is just pushing back the first tensor's device?

Reply:
oops, that should definitely read
```c++
if (res.size() == 0 || tensor.device() != res[0]) {
```

Reply:
But funnily
> res[0] == tensor[0].device is always true"

is too harsh.

```C++
tensor[0].device() != res[0])
```
was (accidentally) correct code. `tensor[0]` is a one-element slice of `tensor`, so `tensor[0].device()` was the same as `tensor.device()`.

Reply:
Oh right, I mistook `tensor[0]` as `tensors[0]`, and didn't realize that `tensor[0].device` is also correct (albeit a bit confusing)